File: app/src/main/java/com/adrcotfas/minibrowser/ui/main/MiniWebChromeClient.java
package com.adrcotfas.minibrowser.ui.main;

import android.webkit.WebChromeClient;
import android.webkit.WebView;

public class MiniWebChromeClient extends WebChromeClient {

    interface Listener {
        void onUrlChanged(String url);
        void onProgressChanged(int progress);
    }

    private final Listener listener;

    MiniWebChromeClient(Listener listener) {
        this.listener = listener;
    }

    @Override
    public void onProgressChanged(WebView view, int progress) {
        listener.onProgressChanged(progress);
        if (progress == 100) {
            listener.onUrlChanged(view.getUrl());
        }
    }
}

File: app/src/main/java/com/adrcotfas/minibrowser/ui/main/MiniWebChromeClient.kt
package com.adrcotfas.minibrowser.ui.main

import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebView

class MiniWebChromeClient internal constructor(private val listener: Listener) : WebChromeClient() {
    internal interface Listener {
        fun onUrlChanged(url: String?)
        fun onProgressChanged(progress: Int)
    }

    override fun onProgressChanged(view: WebView, progress: Int) {
        listener.onProgressChanged(progress)
        if (progress == 100) {
            listener.onUrlChanged(view.url)
        }
    }

    override fun onShowCustomView(view: View?, callback: CustomViewCallback?) {
        super.onShowCustomView(view, callback)
    }
}